File: lib/host/vimeo.c
/* 
* Copyright (C) 2009,2010 Toni Gundogdu.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "host.h"

_host_constants(vimeo, "vimeo.com", "flv"); /* Make hd -> flv. */

_host_re(re_id,     "(?i)clip_id=(.*?)\"");
_host_re(re_title,  "(?i)<caption>(.*?)</caption>");
_host_re(re_sign,   "(?i)<request_signature>(.*?)</");
_host_re(re_exp,    "(?i)<request_signature_expires>(.*?)</");
_host_re(re_hd,     "(?i)<hd_button>(\\d+)</");

QUVIcode
handle_vimeo(const char *url, _quvi_video_t video) {
    char *content, *config_url, *config, *sign, *exp;
    char *format, *quality, *lnk;
    int hd_avail;
    QUVIcode rc;

    hd_avail = 0;

    /* host id */
    _host_id("vimeo");

    /* common */
    rc = parse_page_common(url, video, &content, re_id, NULL);

    _free(content);

    if (rc != QUVI_OK)
        return (rc);

    /* config */
    asprintf(&config_url,
        "http://vimeo.com/moogaloop/load/clip:%s", video->id);

    rc = fetch_to_mem(video, config_url, QUVISTATUSTYPE_CONFIG, &config);

    _free(config_url);

    if (rc != QUVI_OK)
        return (rc);

    /* title */
    _free(video->title);

    rc = regexp_capture(
        video->quvi,
        config,
        re_title,
        0,
        0,
        &video->title,
        (void *) 0
    );

    if (rc != QUVI_OK) {
        _free(config);
        return (rc);
    }

    /* req. signature. */
    rc = regexp_capture(
        video->quvi,
        config,
        re_sign,
        0,
        0,
        &sign,
        (void *) 0
    );

    if (rc == QUVI_OK) {
        /* req. signature expires */
        rc = regexp_capture(
            video->quvi,
            config,
            re_exp,
            0,
            0,
            &exp,
            (void *) 0
        );

        if (rc == QUVI_OK) {
            QUVIcode _rc; /* do not override rc */
            char *hd;

            _rc = regexp_capture(
                video->quvi,
                config,
                re_hd,
                0,
                0,
                &hd,
                (void *) 0
            );

            if (hd && !strcmp(hd, "1"))
                hd_avail = 1;

            _free(hd);
        }
    }

    _free(config);

    if (rc != QUVI_OK) {
        _free(exp);
        _free(sign);
        return (rc);
    }

    format  = video->quvi->format;

    quality = 
        (format && (!strcmp(format,"hd") || !strcmp(format,"best")) && hd_avail)
        ? "hd"
        : "sd";

    /* video link */
    asprintf(&lnk,
        "http://vimeo.com/moogaloop/play/clip:%s/%s/%s/?q=%s",
        video->id, sign, exp, quality);

    _free(sign);
    _free(exp);

    rc = add_video_link(&video->link, "%s", lnk);

    _free(lnk);

    return (QUVI_OK);
}

File: share/lua/website/vimeo.lua

-- quvi
-- Copyright (C) 2010,2011  Toni Gundogdu <legatvs@gmail.com>
--
-- This file is part of quvi <http://quvi.sourceforge.net/>.
--
-- This library is free software; you can redistribute it and/or
-- modify it under the terms of the GNU Lesser General Public
-- License as published by the Free Software Foundation; either
-- version 2.1 of the License, or (at your option) any later version.
--
-- This library is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-- Lesser General Public License for more details.
--
-- You should have received a copy of the GNU Lesser General Public
-- License along with this library; if not, write to the Free Software
-- Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
-- 02110-1301  USA
--

-- w/ HD: <http://vimeo.com/1485507>
-- no HD: <http://vimeo.com/10772672>

local Vimeo = {} -- Utility functions unique to this script.

-- Identify the script.
function ident(self)
    package.path = self.script_dir .. '/?.lua'
    local C      = require 'quvi/const'
    local r      = {}
    r.domain     = "vimeo.com"
    r.formats    = "default|best"
    r.categories = C.proto_http
    local U      = require 'quvi/util'
    r.handles    = U.handles(self.page_url, {r.domain}, {"/%d+$"})
    return r
end

-- Query available formats.
function query_formats(self)
    local config  = Vimeo.get_config(self)
    local formats = Vimeo.iter_formats(self, config)

    local t = {}
    for _,v in pairs(formats) do
        table.insert(t, Vimeo.to_s(v))
    end

    table.sort(t)
    self.formats = table.concat(t, "|")

    return self
end

-- Parse media URL.
function parse(self)
    self.host_id  = "vimeo"
    local config  = Vimeo.get_config(self)

    local _,_,s = config:find("<caption>(.-)</")
    self.title  = s or error("no match: media title")

    local _,_,s = config:find('<duration>(%d+)')
    self.duration = (tonumber(s) or 0) * 1000 -- to msec

    local _,_,s = config:find('<thumbnail>(.-)<')
    self.thumbnail_url = s or ''

    local formats = Vimeo.iter_formats(self, config)
    local U       = require 'quvi/util'
    self.url      = {U.choose_format(self, formats,
                                     Vimeo.choose_best,
                                     Vimeo.choose_default,
                                     Vimeo.to_s).url
                        or error("no match: media url")}
    return self
end

--
-- Utility functions
--

function Vimeo.normalize(url)
    url = url:gsub("player.", "") -- player.vimeo.com
    url = url:gsub("/video/", "/") -- player.vimeo.com
    return url
end

function Vimeo.get_config(self)
    self.page_url = Vimeo.normalize(self.page_url)

    local _,_,s   = self.page_url:find('vimeo.com/(%d+)')
    self.id       = s or error("no match: media id")

    local config_url = "http://vimeo.com/moogaloop/load/clip:" .. self.id
    local config = quvi.fetch(config_url, {fetch_type = 'config'})

    if config:find('<error>') then
        local _,_,s = config:find('<message>(.-)[\n<]')
        error( (not s) and "no match: error message" or s )
    end

    return config
end

function Vimeo.iter_formats(self, config)
    local _,_,s = config:find('<isHD>(%d+)')
    local isHD  = tonumber(s) or 0

    local t = {}
    Vimeo.add_format(self, config, t, 'sd')
    if isHD == 1 then
        Vimeo.add_format(self, config, t, 'hd')
    end

    return t
end

function Vimeo.add_format(self, config, t, quality)
    table.insert(t,
        {quality=quality,
         url=Vimeo.to_url(self, config, quality)})
end

function Vimeo.choose_best(formats) -- Last is 'best'
    local r
    for _,v in pairs(formats) do r = v end
    return r
end

function Vimeo.choose_default(formats) -- First is 'default'
    for _,v in pairs(formats) do return v end
end

function Vimeo.to_url(self, config, quality)
    local _,_,s = config:find("<request_signature>(.-)</")
    local sign  = s or error("no match: request signature")

    local _,_,s = config:find("<request_signature_expires>(.-)</")
    local exp   = s or error("no match: request signature expires")

    local fmt_s = "http://vimeo.com/moogaloop/play/clip:%s/%s/%s/?q=%s"

    return string.format(fmt_s, self.id, sign, exp, quality)
end

function Vimeo.to_s(t)
    return string.format("%s", t.quality)
end

-- vim: set ts=4 sw=4 tw=72 expandtab:
